devito/finite_differences/coefficients.py

```diff
@@ -1,73 +1,15 @@
-from functools import cached_property
+from devito.deprecations import deprecations
 
-import numpy as np
-
-from devito.finite_differences import Weights, generate_indices
-from devito.finite_differences.tools import numeric_weights
-from devito.tools import filter_ordered, as_tuple
-
-__all__ = ['Coefficient', 'Substitutions', 'default_rules']
+__all__ = ['Coefficient', 'Substitutions']
 
 
 class Coefficient:
-    """
-    Prepare custom coefficients to pass to a Substitutions object.
-
-    Parameters
-    ----------
-    deriv_order : int
-        The order of the derivative being taken.
-    function : Function
-        The function for which the supplied coefficients
-        will be used.
-    dimension : Dimension
-        The dimension with respect to which the
-        derivative is being taken.
-    weights : np.ndarray
-        The set of finite difference weights
-        intended to be used in place of the standard
-        weights (obtained from a Taylor expansion).
-
-    Examples
-    --------
-    >>> import numpy as np
-    >>> from devito import Grid, Function, Coefficient
-    >>> grid = Grid(shape=(4, 4))
-    >>> u = Function(name='u', grid=grid, space_order=2, coefficients='symbolic')
-    >>> x, y = grid.dimensions
-
-    Now define some partial d/dx FD coefficients of the Function u:
-
-    >>> u_x_coeffs = Coefficient(1, u, x, np.array([-0.6, 0.1, 0.6]))
-
-    And some partial d^2/dy^2 FD coefficients:
-
-    >>> u_y2_coeffs = Coefficient(2, u, y, np.array([0.0, 0.0, 0.0]))
-
-    """
-
     def __init__(self, deriv_order, function, dimension, weights):
-
-        self._check_input(deriv_order, function, dimension, weights)
-
-        # Ensure the given set of weights is the correct length
-        try:
-            wl = weights.shape[-1]-1
-        except AttributeError:
-            wl = len(weights)-1
-        if dimension.is_Time:
-            if wl != function.time_order:
-                raise ValueError("Number of FD weights provided does not "
-                                 "match the functions space_order")
-        elif dimension.is_Space:
-            if wl != function.space_order:
-                raise ValueError("Number of FD weights provided does not "
-                                 "match the functions space_order")
-
+        deprecations.coeff_warn
```

Comment:
nitpicking: perhaps a memoized_meth would be neater because you're performing an action rather than retrieving a property

but really, this is super-nitpicking. Can revisit another day